### ISSUE TITLE
Implement BCS_P table

### ISSUE THREAD

**Issue by upsonp** [upsonp/dart#106]
The BCS_P table contains header information required to complete the BioChem upload for plankton. The BCS_P table describes the gear type. Robert tells me Phytoplankton are captured in bottles while Zooplankton are captured in RingNets.

I currently have Zooplankton creating and adding to a BCD_P table, but am rewriting some of the BioChem upload code to be more modular. Each samples/[dir]/templates/html/ directory will contain a [dir]_upload.js script that contains code specifically for calling that samples API to get BCD/BCS data. To do this for Zooplankton I'll need the gear type in the BCS table.

**Comment by upsonp** [comment]
Found in the Form_frm_Mission_exports code

```VB
Private Sub btn_BcsBcd_Plankton_Click()
Dim tStr As String, bcsName As String, bcdName As String
Dim skipRec As Boolean
Dim bcsSeq As String, bcdSeq As Double, cBatch As Long
Dim uCreatedDate As Date
Dim azmpBCSrst As Recordset, azmpBCDrst As Recordset, expBCSrst As Recordset, expBCDrst As Recordset
disableMenu Me
'This report relies on multipal queries and tables
On Error GoTo btn_BcsBcd_Plankton_Click_Error
SysCmd acSysCmdInitMeter, "Preparing Plankton BCS, BCD Export...", 100
SysCmd acSysCmdUpdateMeter, 0
Me.Label_Output_Dir.Caption = "Working..." & vbCrLf & "Note: Only selected Plankton data will be sent to Biochem."

If Len(Nz(AZMPDB, "")) = 0 Then
    AZMPDB = ReadAZMPDbLoc
End If
'build new bcs and bcd tables
bcsName = Me.Combo_SelectMission & "_BCS_P"
If TableExists(bcsName) Then DoCmd.DeleteObject acTable, bcsName
CreatePlanktonBCSTbl bcsName
bcdName = Me.Combo_SelectMission & "_BCD_P"
If TableExists(bcdName) Then DoCmd.DeleteObject acTable, bcdName
CreatePlanktonBCDTbl bcdName

'Collect data for BCS
tStr = "SELECT distinct m.*, e.*, h.*, Val(e.COLLECTOR_EVENT_ID) " _
    & "FROM (_MISSIONS AS m INNER JOIN (_EVENTS AS e INNER JOIN _PLANKTNHEDRS AS h ON e.EVENT_SEQ = h.EVENT_SEQ) ON m.MISSION_SEQ = e.MISSION_SEQ) INNER JOIN _PLANKTNGENERALS AS pg ON h.PLANKTON_SEQ = pg.PLANKTON_SEQ IN '" & AZMPDB & "' " _
    & "where m.NAME = '" & Me.Combo_SelectMission & "' " _
    & "Order by val(e.COLLECTOR_EVENT_ID), h.COLLECTOR_SAMPLE_ID;"
Set azmpBCSrst = CurrentDb.OpenRecordset(tStr)
If azmpBCSrst.RecordCount > 0 Then
    azmpBCSrst.MoveLast
    azmpBCSrst.MoveFirst
    SysCmd acSysCmdInitMeter, "Formatting Data Event ", azmpBCSrst.RecordCount
    SysCmd acSysCmdUpdateMeter, azmpBCSrst.AbsolutePosition
    'cBatch
    uCreatedDate = Now()
    Set expBCSrst = CurrentDb.OpenRecordset("Select * from " & bcsName & " order by EVENT_COLLECTOR_EVENT_ID, PL_HEADR_COLLECTOR_SAMPLE_ID;")
    Do While Not azmpBCSrst.EOF
        'writeBCS
        SysCmd acSysCmdInitMeter, "Formatting Data Event " & azmpBCSrst.Fields("COLLECTOR_EVENT_ID"), azmpBCSrst.RecordCount
        SysCmd acSysCmdUpdateMeter, azmpBCSrst.AbsolutePosition
        bcsSeq = azmpBCSrst.Fields("Descriptor") & "_" & azmpBCSrst.Fields("COLLECTOR_EVENT_ID") & "_" & IIf(IsNull(azmpBCSrst.Fields("COLLECTOR_SAMPLE_ID")), "0", azmpBCSrst.Fields("COLLECTOR_SAMPLE_ID")) & "_" & azmpBCSrst.Fields("Gear_seq")
        'If bcsSeq = 79 Then Stop
        DoEvents
        If writeBCSP(azmpBCSrst, expBCSrst, bcsSeq, cBatch, uCreatedDate) Then
            'If sample_Id exists
            If Not IsNull(azmpBCSrst.Fields("COLLECTOR_SAMPLE_ID")) Then
                'collect BCD data for current Sampleid

                tStr = "SELECT m.NAME, m.DESCRIPTOR, e.COLLECTOR_EVENT_ID, e.COLLECTOR_STATION_NAME, h.COLLECTOR_SAMPLE_ID, g.*, d.*, f.*, i.* " _
                    & "FROM (_MISSIONS AS m LEFT JOIN (_EVENTS AS e LEFT JOIN _PLANKTNHEDRS AS h ON e.EVENT_SEQ = h.EVENT_SEQ) ON " _
                    & "m.MISSION_SEQ = e.MISSION_SEQ) LEFT JOIN (((_PLANKTNGENERALS AS g LEFT JOIN _PLANKTNDTAILS AS d ON " _
                    & "g.PLANKTON_GENERAL_SEQ = d.PLANKTON_GENERAL_SEQ) LEFT JOIN _PLANKTNFREQS AS f ON " _
                    & "g.PLANKTON_GENERAL_SEQ = f.PLANKTON_GENERAL_SEQ) LEFT JOIN _PLANKTNINDIVIDLS AS i ON " _
                    & "g.PLANKTON_GENERAL_SEQ = i.PLANKTON_GENERAL_SEQ) ON h.PLANKTON_SEQ = g.PLANKTON_SEQ " _
                    & "IN '" & AZMPDB & "' " _
                    & "WHERE (((g.PLANKTON_SEQ) = " & azmpBCSrst.Fields("PLANKTON_SEQ") & ")) " _
                    & "ORDER BY g.PLANKTON_GENERAL_SEQ, d.PLANKTON_DETAIL_SEQ, f.PLANKTON_FREQUENCY_SEQ, i.PLANKTON_INDIVIDUAL_SEQ;"

                Set azmpBCDrst = CurrentDb.OpenRecordset(tStr)
                If azmpBCDrst.RecordCount > 0 Then
                    azmpBCDrst.MoveLast: azmpBCDrst.MoveFirst
                    Do While Not azmpBCDrst.EOF
                        bcdSeq = GetMaxPk("PLANK_DATA_NUM", bcdName)
                        bcdSeq = bcdSeq + 1
                        Set expBCDrst = CurrentDb.OpenRecordset("Select * from " & bcdName & " order by PLANK_DATA_NUM;")
                        skipRec = False
                        DoEvents
                        If Not writeBCDP(azmpBCDrst, expBCDrst, bcsSeq, bcdSeq, cBatch, skipRec, uCreatedDate) Then
                            'Build an error table
                        End If
                        If skipRec Then
                            azmpBCDrst.MoveNext
                            skipRec = False
                        End If
                        If Not azmpBCDrst.EOF Then
                            azmpBCDrst.MoveNext
                        End If
                    Loop
                    azmpBCDrst.Close
                End If
            End If
        Else
            
            'Build an error table
            MsgBox "Not all data was exported. Error writing to BCS table."
            azmpBCSrst.MoveLast
        End If
        azmpBCSrst.MoveNext
    Loop
    SysCmd acSysCmdInitMeter, "Exporting Tables ", 100
    writeTable2CSV Me, Me.Label_Output_Dir, bcsName, True
    writeTable2CSV Me, Me.Label_Output_Dir, bcdName, True
End If

azmpBCSrst.Close

'********************************************
enableMenu Me
SysCmd acSysCmdClearStatus
Exit Sub
btn_BcsBcd_Plankton_Click_Error:
MsgBox "Problems were encountered during the export operation. Export aborted.", vbExclamation, "Export error"
Resume
Me.Label_Output_Dir.Caption = "Aborted..."
SysCmd acSysCmdClearStatus
If Not azmpBCSrst Is Nothing Then
    azmpBCSrst.Close
    Set azmpBCSrst = Nothing
End If
If Not azmpBCDrst Is Nothing Then
    'azmpBCDrst.Close
    Set azmpBCDrst = Nothing
End If
On Error GoTo 0
enableMenu Me
End Sub
```

**Comment by upsonp** [comment]
This is how the 'plank_sample_key_value' is constructed:

```VB
        bcsSeq = azmpBCSrst.Fields("Descriptor") & "_" & azmpBCSrst.Fields("COLLECTOR_EVENT_ID") & "_" & IIf(IsNull(azmpBCSrst.Fields("COLLECTOR_SAMPLE_ID")), "0", azmpBCSrst.Fields("COLLECTOR_SAMPLE_ID")) & "_" & azmpBCSrst.Fields("Gear_seq")
```

**Comment by upsonp** [comment]
This is how the azmpBCSrst query is constructed:

```VB
tStr = "SELECT distinct m.*, e.*, h.*, Val(e.COLLECTOR_EVENT_ID) " _
    & "FROM (_MISSIONS AS m INNER JOIN (_EVENTS AS e INNER JOIN _PLANKTNHEDRS AS h ON e.EVENT_SEQ = h.EVENT_SEQ) ON m.MISSION_SEQ = e.MISSION_SEQ) INNER JOIN _PLANKTNGENERALS AS pg ON h.PLANKTON_SEQ = pg.PLANKTON_SEQ IN '" & AZMPDB & "' " _
    & "where m.NAME = '" & Me.Combo_SelectMission & "' " _
    & "Order by val(e.COLLECTOR_EVENT_ID), h.COLLECTOR_SAMPLE_ID;"
```

**Comment by upsonp** [comment]
This is the function where the _PLANKTNHEDRS DB is written to. This is where Gear Sequence is set.

```VB
Function writeAzmpPlkHead(ByRef writeToRst, ByRef writeFromRst, writeSeq As Double, linkToSeq As Double, ActivitySeq As Double, addMode As Boolean, updated As Boolean, moveTo As String, cBatch As Long) As Boolean

Dim dbs As Database, tStr As String
writeAzmpPlkHead = False
updated = False
On Error GoTo writeAzmpHeadErr:
    If addMode Then
        If moveTo = "AZMP" Then
            writeToRst.Fields("Plankton_SEQ") = writeSeq
            writeToRst.Fields("EVENT_SEQ") = linkToSeq
            writeToRst.Fields("ACTIVITY_SEQ") = ActivitySeq
        Else
            writeToRst.Fields("Pla_HEADR_EDT_SEQ") = writeSeq
            writeToRst.Fields("EVENT_EDT_SEQ") = linkToSeq
            writeToRst.Fields("ACTIVITY_EDT_SEQ") = ActivitySeq
        End If
        updated = True
    End If
    'DATA_CENTER_CODE
    'If checkAudit(writeToRst, writeFromRst, "DATA_CENTER_CODE", "DATA_CENTER_CODE", "_DISCRETEHEDRS", moveTo) Then
    '    updated = True
    'End If
    
    'GEAR_SEQ
    If checkAudit(writeToRst, writeFromRst, "GEAR_SEQ", "GEAR_SEQ", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
   
    'START_DATE
    If checkAudit(writeToRst, writeFromRst, "START_DATE", "START_DATE", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'END_DATE
    If checkAudit(writeToRst, writeFromRst, "END_DATE", "END_DATE", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'START_TIME
    If checkAudit(writeToRst, writeFromRst, "START_TIME", "START_TIME", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'END_TIME
    If checkAudit(writeToRst, writeFromRst, "END_TIME", "END_TIME", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'TIME_QC_CODE
    If checkAudit(writeToRst, writeFromRst, "TIME_QC_CODE", "TIME_QC_CODE", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    
    'START_LAT
    If checkAudit(writeToRst, writeFromRst, "START_LAT", "START_LAT", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'END_LAT
    If checkAudit(writeToRst, writeFromRst, "END_LAT", "END_LAT", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'START_LON
    If checkAudit(writeToRst, writeFromRst, "START_LON", "START_LON", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'END_LON
    If checkAudit(writeToRst, writeFromRst, "END_LON", "END_LON", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
        
   
    'POSITION_QC_CODE
    If checkAudit(writeToRst, writeFromRst, "POSITION_QC_CODE", "POSITION_QC_CODE", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'START_DEPTH
    If checkAudit(writeToRst, writeFromRst, "START_DEPTH", "START_DEPTH", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'END_DEPTH
    If checkAudit(writeToRst, writeFromRst, "END_DEPTH", "END_DEPTH", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'SOUNDING
    If checkAudit(writeToRst, writeFromRst, "SOUNDING", "SOUNDING", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'VOLUME
    If checkAudit(writeToRst, writeFromRst, "VOLUME", "VOLUME", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'VOLUME_METHOD_SEQ
    If checkAudit(writeToRst, writeFromRst, "VOLUME_METHOD_SEQ", "VOLUME_METHOD_SEQ", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'LARGE_PLANKTON_REMOVED
    If checkAudit(writeToRst, writeFromRst, "LARGE_PLANKTON_REMOVED", "LARGE_PLANKTON_REMOVED", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    'MESH_SIZE
    If checkAudit(writeToRst, writeFromRst, "MESH_SIZE", "MESH_SIZE", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    'COLLECTION_METHOD_SEQ
    If checkAudit(writeToRst, writeFromRst, "COLLECTION_METHOD_SEQ", "COLLECTION_METHOD_SEQ", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    'COLLECTOR_DEPLOYMENT_ID
    If checkAudit(writeToRst, writeFromRst, "COLLECTOR_DEPLOYMENT_ID", "COLLECTOR_DEPLOYMENT_ID", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'COLLECTOR_SAMPLE_ID
    If checkAudit(writeToRst, writeFromRst, "COLLECTOR_SAMPLE_ID", "COLLECTOR_SAMPLE_ID", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'PROCEDURE_SEQ
    If checkAudit(writeToRst, writeFromRst, "PROCEDURE_SEQ", "PROCEDURE_SEQ", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    'PRESERVATION_SEQ
    If checkAudit(writeToRst, writeFromRst, "PRESERVATION_SEQ", "PRESERVATION_SEQ", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    'STORAGE_SEQ
    If checkAudit(writeToRst, writeFromRst, "STORAGE_SEQ", "STORAGE_SEQ", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    'COLLECTOR
    If checkAudit(writeToRst, writeFromRst, "COLLECTOR", "COLLECTOR", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'COLLECTOR_COMMENT
    If checkAudit(writeToRst, writeFromRst, "COLLECTOR_COMMENT", "COLLECTOR_COMMENT", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'METERS_SQD_FLAG
    If checkAudit(writeToRst, writeFromRst, "METERS_SQD_FLAG", "METERS_SQD_FLAG", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'DATA_MANAGER_COMMENT
    If checkAudit(writeToRst, writeFromRst, "DATA_MANAGER_COMMENT", "DATA_MANAGER_COMMENT", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
   'RESPONSIBLE_GROUP
   If checkAudit(writeToRst, writeFromRst, "RESPONSIBLE_GROUP", "RESPONSIBLE_GROUP", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    
    'SHARED_DATA
    If checkAudit(writeToRst, writeFromRst, "SHARED_DATA", "SHARED_DATA", "_PLANKTNHEDRS", moveTo) Then
        updated = True
    End If
    writeAzmpPlkHead = True
Exit 
```

**Comment by upsonp** [comment]
Found this in the ElogEvent code in the AZMP template. Seems to be where Gear Sequence and collection Method Sequence are set.

```VB
Case "RINGNET"
    If cMesh = 202 Then
        cInstrumentSeq = 90000102
    ElseIf cMesh = 76 Then
        cInstrumentSeq = 90000105
    ElseIf cMesh = 70 Then
        cInstrumentSeq = 90000105
    End If
    COLLECTION_METHOD_SEQ = 90000001
    VOLUME_METHOD_SEQ = 90000004
    
    Volumn = Vol_Ideal
Case "RING NET"
    If cMesh = 202 Then
        cInstrumentSeq = 90000102
    ElseIf cMesh = 76 Then
        cInstrumentSeq = 90000105
    ElseIf cMesh = 70 Then
        cInstrumentSeq = 90000105
    End If
    COLLECTION_METHOD_SEQ = 90000001
    VOLUME_METHOD_SEQ = 90000004
    
    Volumn = Vol_Ideal
Case "VERTICAL NET"
    If cMesh = 202 Then
        cInstrumentSeq = 90000102
    ElseIf cMesh = 76 Then
        cInstrumentSeq = 90000105
    ElseIf cMesh = 70 Then
        cInstrumentSeq = 90000105
    End If
    COLLECTION_METHOD_SEQ = 90000001
    VOLUME_METHOD_SEQ = 90000004
    
    Volumn = Vol_Ideal
Case "BIONESS"
    cInstrumentSeq = 90000078
    COLLECTION_METHOD_SEQ = 90000012
Case "MULTINET"
    cInstrumentSeq = 90000093
Case "CTD"
    cInstrumentSeq = 90000019
Case "MANTA"
    cInstrumentSeq = 90000214
    COLLECTION_METHOD_SEQ = 90000020
    VOLUME_METHOD_SEQ = 90000002
Case Else
    cInstrumentSeq = 0
    cMesh = 0
    COLLECTION_METHOD_SEQ = 0
    COLLECTION_METHOD_SEQ = 0
End Select
```

**Comment by upsonp** [comment]
This is in the Phyto code for the AZMP template. Seems to be where a BCS_P entry is created for Phytoplankton.

```VB
Function AddPlHeader(newPlankHdrSeq As Double, newActivitySeq As Double, newEventSeq As Double, cSampleID As String, plHdrRst As Recordset, discreteHdrRst As Recordset, cCollector As String, cBatch) As Boolean
AddPlHeader = False
newPlankHdrSeq = GetMaxPk("PLA_HEADR_EDT_SEQ", "_PLANKTNHEDREDTS")
If newPlankHdrSeq = 0 Then
    newPlankHdrSeq = 1
Else
    newPlankHdrSeq = newPlankHdrSeq + 1
End If
    With plHdrRst
        .AddNew
        .Fields("PLA_HEADR_EDT_SEQ") = newPlankHdrSeq
        .Fields("DATA_CENTER_CODE") = discreteHdrRst.Fields("DATA_CENTER_CODE")
        .Fields("START_DATE") = discreteHdrRst.Fields("sDate")
        .Fields("GEAR_SEQ") = 90000019 'Bottle
        .Fields("END_DATE") = discreteHdrRst.Fields("eDate")
        .Fields("START_TIME") = discreteHdrRst.Fields("sTime")
        .Fields("END_TIME") = discreteHdrRst.Fields("eTime")

        .Fields("PHASE_OF_DAYLIGHT") = ""
        .Fields("START_LAT") = discreteHdrRst.Fields("sLat")
        .Fields("END_LAT") = discreteHdrRst.Fields("eLat")
        .Fields("START_LON") = discreteHdrRst.Fields("sLon")
        .Fields("END_LON") = discreteHdrRst.Fields("eLon")
        .Fields("TIME_QC_CODE") = 1
        .Fields("POSITION_QC_CODE") = 1
        .Fields("SOUNDING") = discreteHdrRst.Fields("Sounding")
        .Fields("MESH_SIZE") = 0
        .Fields("COLLECTOR_SAMPLE_ID") = cSampleID
        .Fields("COLLECTOR") = cCollector
        
        .Fields("START_DEPTH") = discreteHdrRst.Fields("End_Depth")
        .Fields("END_DEPTH") = 0
        .Fields("Volume") = 0.001
        .Fields("VOLUME_METHOD_SEQ") = 90000010
        
        .Fields("LARGE_PLANKTON_REMOVED") = "N"
        .Fields("METERS_SQD_FLAG") = "Y"
        .Fields("SHARED_DATA") = "N"
        .Fields("COLLECTION_METHOD_SEQ") = 90000010
        .Fields("Procedure_seq") = 90000001
        .Fields("Preservation_Seq") = 90000025
        .Fields("Storage_Seq") = 90000008
        .Fields("COLLECTOR_COMMENT") = "Phytoplankton"
        .Fields("DATA_MANAGER_COMMENT") = "Created using AZMP Data Entry Template"
        .Fields("RESPONSIBLE_GROUP") = "AZMP"
        .Fields("CREATED_BY") = UserName
        .Fields("CREATED_DATE") = Now()
        .Fields("ACTIVITY_EDT_SEQ") = newActivitySeq
        .Fields("_PLANKTNHEDREDTS.EVENT_EDT_SEQ") = newEventSeq
        .Fields("BATCH_SEQ") = cBatch 'Batch
        .Update
        AddPlHeader = True
    End With
End Function
```

**Comment by upsonp** [comment]
The BCD and BCS Plankton tables are included in the #107 merge for zooplankton
